react/base.yml

```diff
@@ -0,0 +1,57 @@
+# Enforce best practices and protect against common mistakes in JSX and React.
+#
+# This is an alternative to openstax-poland/base which includes additional rules
+# from eslint-plugin-react.
+
+extends:
+  - ../base.yml
+  - plugin:react/recommended
+
+settings:
+  react:
+    version: 16.2.0
+
+rules:
+  # Forbid use of .bind() and arrow functions as JSX callbacks, as they are not
+  # comparable and can cause unnecessary re-renders.
+  react/jsx-no-bind: error
+
+  # Forbid uses of fragments with a single child element.
+  react/jsx-no-useless-fragment: error
+
+  # Don't access this.state within this.setState.
+  react/no-access-state-in-setstate: error
```

Comment:
Why is that?
```
const toggleState = () => {
  this.setState({  isOpen: !this.state.isOpen })
}
```

Reply:
Because changes to state aren't executed immediately but scheduled for later, and done in batch before next re-render. This means that this code:

```js
this.setState({ counter: this.state.counter + 1 })
this.setState({ counter: this.state.counter + 1 })
```

will only increment `this.state.counter` by one, not two.

You can see longer explanation in [react/no-access-state-in-setstate's documentation](https://github.com/yannickcr/eslint-plugin-react/blob/master/docs/rules/no-access-state-in-setstate.md).

Reply:
I still don't like it since I can't toggle state for `isOpen` without using callback and that's actually something which I tend to use pretty often

Reply:
But then you have no guarantee that it's actually updated correctly. And even React's own documentation plainly says that you shouldn't do this: https://reactjs.org/docs/state-and-lifecycle.html#state-updates-may-be-asynchronous.

Also, is

```js
this.setState(({ isOpen }) => ({ isOpen: !isOpen }))
```

really that much longer? Plus you can always do

```js
function toggle(flag) {
    return state => { [flag]: !state[flag] }
}

this.setState(toggle('isOpen'))
```

Reply:
They say that you should not relay on this, but it's obviously working in many cases.
I can agree to this rule if it will be set to `warn`